File: core/checklist.py
"""§F3 first-run progress — derived from the fleet, not a topic.

Home owns the checklist until every applicable item is done. Applicability:
enroll first Target; connect Cloudflare unless the first Site is mesh_only;
plant Origin-CA only once a proxied public Site exists (skip mesh-only and
unproxied-until-refused); add a Project (the POST that binds dns_zone and
primary_target). `site-dns-unbound` is not a state this can observe.
"""

ITEM_IDS = (
    "enroll_target",
    "connect_cloudflare",
    "plant_origin_ca",
    "add_project",
)


def first_run_progress(workspace=None):
    from core.models import DnsAccount, Project, Site, Target

    sites = Site.objects.order_by("pk")
    targets = Target.objects.all()
    accounts = DnsAccount.objects.all()
    projects = Project.objects.all()
    if workspace is not None:
        sites = sites.filter(project__workspace=workspace)
        targets = targets.filter(zone__workspace=workspace)
        accounts = accounts.filter(workspace=workspace)
        projects = projects.filter(workspace=workspace)
    first = sites.first()
    mesh_only_first = (
        first is not None and first.exposure == Site.Exposure.MESH_ONLY
    )
    has_proxied_public = sites.filter(
        exposure=Site.Exposure.PUBLIC, proxied=True,
    ).exists()
    items = [
        {
            "id": "enroll_target",
            "applicable": True,
            "done": targets.exists(),
        },
        {
            "id": "connect_cloudflare",
            "applicable": not mesh_only_first,
            "done": accounts.exists(),
        },
        {
            "id": "plant_origin_ca",
            "applicable": (not mesh_only_first) and has_proxied_public,
            "done": accounts.exclude(origin_ca_key_ref="").exists(),
        },
        {
            "id": "add_project",
            "applicable": True,
            "done": projects.exists(),
        },
    ]
    return {
        "owns_home": any(row["applicable"] and not row["done"] for row in items),
        "items": items,
    }

File: frontend/src/checklist.js
// §F3 first-run card copy + derived remaining items. Progress itself comes
// from GET /api/v1/first-run/ (server-derived). This module does not invent
// a topic — Home refreshes on the existing `findings` client.

export const CHECKLIST_COPY = {
  enroll_target: {
    sentence: "Enroll the first target — provision a machine and it appears in the fleet.",
    button: "Copy the provision command",
  },
  connect_cloudflare: {
    sentence: "Connect Cloudflare — paste a single-zone token so public sites can bind a zone.",
    button: "Connect Cloudflare",
  },
  plant_origin_ca: {
    sentence: "Plant the Origin-CA key from a Hub-local file — required before the first proxied public deploy.",
    button: "Plant Origin-CA",
  },
  add_project: {
    sentence: "Add a project — the create that binds the DNS zone and primary target.",
    button: "Add a project",
  },
};

export function remainingItems(progress) {
  return (progress?.items || []).filter((item) => item.applicable && !item.done);
}

export function checklistOwnsHome(progress) {
  return Boolean(progress?.owns_home);
}
